### PR TITLE
install: use 4k image when when targeting 4k drives

### PR DESCRIPTION
**Pull request by jlebon**
Detect the sector size of the target device, then select the right
format from the stream metadata based on that.

Closes: #201

### CODE REVIEW

**Comment by jlebon** [comment]
(Working on enhancing `kola testiso` to also test 4k installs.)

File: src/cmdline.rs [hunk 1]
```diff
@@ -363,13 +371,19 @@ fn parse_install(matches: &ArgMatches) -> Result<Config> {
         } else {
             None
         };
+        let format = match get_sector_size_for_path(Path::new(&device))?.get() {
+            4096 => "4k.raw.xz",
+            // could bail on non-512, but let's be optimistic and just try the regular 512b image
+            _ => "raw.xz",
```
Comment (cgwalters) [inline src/cmdline.rs:377]:
```suggestion
            512 => "raw.xz",
            format => { eprintln!("Warning: Unhandled sector size '{}'; assuming 512", format); "raw.xz" },
```

File: src/blockdev.rs [hunk 2]
```diff
@@ -209,6 +210,19 @@ pub fn reread_partition_table(file: &mut File) -> Result<()> {
     Ok(())
 }
 
+pub fn get_sector_size_for_path(device: &Path) -> Result<NonZeroU32> {
+    let dev = OpenOptions::new()
+        .read(true)
+        .open(Path::new(device))
+        .chain_err(|| format!("opening {:?}", device))?;
+
+    if !dev.metadata()?.file_type().is_block_device() {
```
Comment (bgilbert) [inline src/blockdev.rs:219]: Missing `chain_err` on `metadata()`.

File: src/cmdline.rs [hunk 3]
```diff
@@ -12,10 +12,13 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::path::Path;
```
Comment (bgilbert) [inline src/cmdline.rs:15]: The codebase generally groups `std` with other imports.

File: src/cmdline.rs [hunk 4]
```diff
@@ -363,13 +371,19 @@ fn parse_install(matches: &ArgMatches) -> Result<Config> {
         } else {
             None
         };
+        let format = match get_sector_size_for_path(Path::new(&device))?.get() {
```
Comment (bgilbert) [inline src/cmdline.rs:374]: `chain_err`?

**Comment by jlebon** [comment]
OK so https://github.com/coreos/coreos-assembler/pull/1312 enhances `kola testiso` for testing metal4k image installs. Will hook it up here and in the pipeline too afterwards, though it doesn't actually test the new codepath here since we're not doing a stream install. We could work on mocking that though... not sure if we should block on this. I *did* test this code on the current stable stream metadata though and it did the right thing. :)

Will address review comments tomorrow!

File: src/cmdline.rs (hunk 3 above)
Comment (jlebon) [reply src/cmdline.rs]: Thanks, fixed! (Though I personally prefer separating std imports.)

**Comment by jlebon** [comment]
I added another commit here which adds metal4k testing in CI!

**Pull request merged**